src/Microsoft.Identity.Client/Mats/Mats.cs

```diff
@@ -154,6 +154,39 @@ public void EndAction(IActionHandle action, AuthOutcome outcome, ErrorSource err
             _actionStore.EndAdalAction((AdalAction)action, outcome, errorSource, error, errorDescription);
             _uploader.Upload(GetEventsForUpload());
         }
+
+        public void EndAction(IActionHandle action, Exception ex)
+        {
+            AuthOutcome outcome = AuthOutcome.Failed;
+            ErrorSource errorSource = ErrorSource.AuthSdk;
+            string error = ex.Message;
+            string errorDescription = ex.Message;
+
+            switch (ex)
+            {
+            case MsalUiRequiredException uiEx:
+                errorSource = ErrorSource.Service;
+                error = uiEx.ErrorCode;
+                break;
+            case MsalServiceException svcEx:
+                errorSource = ErrorSource.Service;
+                error = svcEx.ErrorCode;
+                break;
+            case MsalClientException cliEx:
+                errorSource = ErrorSource.Client;
+                error = cliEx.ErrorCode;
+                break;
+            case MsalException msalEx:
+                errorSource = ErrorSource.AuthSdk;
+                error = msalEx.ErrorCode;
+                break;
+            default:
+                errorSource = ErrorSource.AuthSdk;
```

Comment:
In this case, maybe the error could be the exception type?

Reply:
done